docs/proposals/component-registration.md

```diff
@@ -0,0 +1,317 @@
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own namespaced API resource (provisionally called **ExternalServiceTarget**). Endpoints Controller will then continue to manage the lifecycle (create/update/delete) of Internal Endpoints based on a selection of Pod instances, but will also do the same for External Endpoints based on a selection of ExternalServiceTargets defined in ServiceSpec.TargetSelector. Endpoints then becomes a (conceptually) read-only API for the user. Endpoints can't actually be read-only, because the Endpoints Controller needs access, but write access may eventually be limited to admin/operator-only.
+
+
+### API Changes
+
+1. Add new namespaced API resource: ExternalServiceTarget
+  - See [Alternate Naming](#alternate-naming)
+  - Similar to Pods
+    - Listable, like most other top-level API resources
+    - Have both Spec and Status (filterable on both/either)
+    - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+    - Describes a single instance
+    - Can be selected by labels, metadata, spec, and status
+    - Is namespaced (components would use a specific namespace, likely "kube-system")
+    - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+    - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+  - Similar to Endpoints (current External Endpoints usage)
+    - Spec would have address (ip or host) and ports (optionally named) to connect with an external service provider instance
+    - Manually added for non-component External Service provider instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
+  - Analogous to the ServiceSpec.Selector-Pod relationship
+3. Add a new ExternalServiceTarget Controller that handles probing ExternalServiceTargets
+  - Probe results update ExternalServiceTarget.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceTarget Endpoints
+  - Readiness from ExternalServiceTarget.Status.Conditions analogous to PodStatus.Conditions
+5. Modify components to self-register ExternalServiceTargets (one per component instance) instead of an Endpoint
```

Comment:
right. Should be "modify primary components", because they're all currently external. But many of the controllers could be extracted and deployed as pods after this PR. I'm not sure if the scheduler can... it'd need something to schedule it.